File: firmware/esp32/signer.cpp
// firmware/esp32/signer.cpp
// Why sign bytes and not a struct or a JSON object:
//  - A C struct's layout is padding- and endianness-dependent. The same logical report
//    signed on the ESP32 and verified on x86 would hash differently.
//  - A dict/JSON object has no inherent key order, and encoders differ on whitespace,
//    integer formatting and escaping. Two encoders produce two different signatures for
//    the same data, so verification becomes a coin toss.
// So the tuple is flattened to one explicit byte string: a version tag, then each field
// length-prefixed and big-endian. Any change to that layout must bump kFormatVersion,
// because the gateway builds the identical bytes in gateway/signer.py.
//
// The seed is read from NVS at runtime. It is never compiled into the image.
#include "signer.h"

#include <cstring>

#include "esp_log.h"
#include "mbedtls/ctr_drbg.h"
#include "mbedtls/entropy.h"
#include "sodium/crypto_sign_ed25519.h"

namespace meter {

static const char* TAG = "signer";

// Bump this if the encoding below changes in any way.
static constexpr uint8_t kFormatVersion = 1;
static constexpr char kDomain[] = "ble-usage-meter/report/v1";

static void put_u64_be(std::vector<uint8_t>& out, uint64_t v) {
  for (int i = 7; i >= 0; --i) out.push_back(static_cast<uint8_t>((v >> (i * 8)) & 0xFF));
}

static void put_lp_string(std::vector<uint8_t>& out, const std::string& s) {
  // 2-byte big-endian length, then raw UTF-8 bytes. Length prefixing stops
  // ("ab","c") and ("a","bc") producing the same serialisation.
  uint16_t n = static_cast<uint16_t>(s.size());
  out.push_back(static_cast<uint8_t>((n >> 8) & 0xFF));
  out.push_back(static_cast<uint8_t>(n & 0xFF));
  out.insert(out.end(), s.begin(), s.end());
}

std::vector<uint8_t> canonical_bytes(const ReportTuple& r) {
  std::vector<uint8_t> out;
  out.reserve(96 + r.device_id.size());

  // Domain separation: a signature over a report can never be replayed as a signature
  // over some other message this key might sign later.
  out.insert(out.end(), kDomain, kDomain + sizeof(kDomain) - 1);
  out.push_back(kFormatVersion);

  put_lp_string(out, r.device_id);
  put_u64_be(out, r.count);
  put_u64_be(out, r.sequence);
  put_u64_be(out, r.window_start);
  put_u64_be(out, r.window_end);
  return out;
}

bool Signer::load_seed(const uint8_t seed[32]) {
  if (seed == nullptr) return false;
  std::memcpy(seed_, seed, 32);

  uint8_t sk[crypto_sign_ed25519_SECRETKEYBYTES];
  if (crypto_sign_ed25519_seed_keypair(pub_, sk, seed_) != 0) {
    ESP_LOGE(TAG, "keypair derivation failed");
    std::memset(seed_, 0, sizeof(seed_));
    return false;
  }
  std::memset(sk, 0, sizeof(sk));
  ready_ = true;
  return true;
}

bool Signer::public_key(uint8_t out[32]) const {
  if (!ready_) return false;
  std::memcpy(out, pub_, 32);
  return true;
}

bool Signer::sign(const ReportTuple& r, uint8_t sig_out[64]) const {
  if (!ready_) return false;

  uint8_t sk[crypto_sign_ed25519_SECRETKEYBYTES];
  uint8_t pk[crypto_sign_ed25519_PUBLICKEYBYTES];
  if (crypto_sign_ed25519_seed_keypair(pk, sk, seed_) != 0) return false;

  std::vector<uint8_t> msg = canonical_bytes(r);
  unsigned long long siglen = 0;
  int rc = crypto_sign_ed25519_detached(sig_out, &siglen, msg.data(), msg.size(), sk);

  std::memset(sk, 0, sizeof(sk));
  return rc == 0 && siglen == 64;
}

}  // namespace meter

File: gateway/signer.py
"""gateway/signer.py

Ed25519 signing over the same canonical byte serialisation the firmware uses, so either
the ESP32 or this gateway can produce a report the backend verifies with one code path.

Why bytes and not a dict:
  - A Python dict has no guaranteed order across producers, and JSON encoders differ on
    whitespace, key order, integer and float formatting, and escaping. Two encoders give
    two signatures for the same data, so verification becomes luck.
  - A C struct is worse: padding and endianness are platform-dependent.
So the tuple is flattened to one explicit byte string, matching signer.cpp exactly: domain
tag, format version, then each field length-prefixed or big-endian fixed width. Any change
here must be mirrored in firmware/esp32/signer.cpp and must bump FORMAT_VERSION.

The private key is read from a file path at runtime. It is never inline in this repo and
never committed. Config ships REPLACE_ME placeholders only.
"""

from __future__ import annotations

import struct
from dataclasses import dataclass
from pathlib import Path

from cryptography.hazmat.primitives import serialization
from cryptography.hazmat.primitives.asymmetric.ed25519 import (
    Ed25519PrivateKey,
    Ed25519PublicKey,
)

DOMAIN = b"ble-usage-meter/report/v1"
FORMAT_VERSION = 1


@dataclass(frozen=True)
class ReportTuple:
    device_id: str
    count: int
    sequence: int
    window_start: int
    window_end: int


def canonical_bytes(r: ReportTuple) -> bytes:
    """Byte-for-byte identical to canonical_bytes() in firmware/esp32/signer.cpp."""
    dev = r.device_id.encode("utf-8")
    if len(dev) > 0xFFFF:
        raise ValueError("device_id too long")
    return b"".join(
        (
            DOMAIN,
            bytes([FORMAT_VERSION]),
            # 2-byte big-endian length prefix stops ("ab","c") and ("a","bc") colliding.
            struct.pack(">H", len(dev)),
            dev,
            struct.pack(">QQQQ", r.count, r.sequence, r.window_start, r.window_end),
        )
    )


class Signer:
    def __init__(self, key: Ed25519PrivateKey) -> None:
        self._key = key

    @classmethod
    def from_file(cls, path: str | Path, password: bytes | None = None) -> "Signer":
        p = Path(path).expanduser()
        if not p.exists():
            raise FileNotFoundError(f"no signing key at {p}, generate one with tools/keygen.py")
        data = p.read_bytes()
        if b"REPLACE_ME" in data:
            raise ValueError(f"{p} is still a placeholder, provision a real key")
        key = serialization.load_pem_private_key(data, password=password)
        if not isinstance(key, Ed25519PrivateKey):
            raise TypeError("signing key must be Ed25519")
        return cls(key)

    def public_key_bytes(self) -> bytes:
        return self._key.public_key().public_bytes(
            encoding=serialization.Encoding.Raw,
            format=serialization.PublicFormat.Raw,
        )

    def sign(self, r: ReportTuple) -> bytes:
        """64-byte detached signature over canonical_bytes(r)."""
        return self._key.sign(canonical_bytes(r))


def verify(public_key: bytes | Ed25519PublicKey, r: ReportTuple, signature: bytes) -> bool:
    """Local self-check. The backend runs the authoritative check in backend/verify.py."""
    pk = (
        public_key
        if isinstance(public_key, Ed25519PublicKey)
        else Ed25519PublicKey.from_public_bytes(public_key)
    )
    try:
        pk.verify(signature, canonical_bytes(r))
    except Exception:
        return False
    return True
